Update Python documentation

swgillespie opened the pull request:
Update the documentation to be clear that Pulumi >=0.16.4 supports Python 3 only. Also replaces the `virtualenv` documentation with `pipenv`, which is a solid tool for managing and installing dependencies.

ellismg commented:
LGTM.  Don't know if you want to mention about that the `python` binary on your path needs to be python 3.6+ or you need to set `PULUMI_PYTHON_CMD` to something like `python3`, or if we just plan to fix that tomorrow.

swgillespie commented:
> Don't know if you want to mention about that the python binary on your path needs to be python 3.6+ or you need to set PULUMI_PYTHON_CMD to something like python3, or if we just plan to fix that tomorrow.

Hmm. I imagine we won't be doing another CLI release in the next few days, so I think it makes sense to document it here. I'll add it.

swgillespie commented:
@ellismg Mind taking a look again at the new text I added in the latest commit? Once https://github.com/pulumi/pulumi/pull/2204 lands I imagine this will be moot but until then I added a note about using `PULUMI_PYTHON_CMD` to force using `python3`.

lukehoban commented on reference/python.md:
```diff
@@ -17,16 +20,23 @@ This will leave behind a `Pulumi.yaml` file, containing some minimal metadata ab
-$ virtualenv .
-$ . ./bin/activate
+$ pipenv --python 3
+$ pipenv install
```
When I follow these instructions myself - I see this:

```
$ pipenv --python 3
Creating a virtualenv for this project...
Pipfile: /Users/luke/go/src/github.com/pulumi/examples/aws-py-stepfunctions/Pipfile
Using /usr/local/bin/python3 (3.7.0) to create virtualenv...
⠋Running virtualenv with interpreter /usr/local/bin/python3
Using base prefix '/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7'
/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/virtualenv.py:1041: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
New python executable in /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B/bin/python3.7
Also creating executable in /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B/bin/python
Installing setuptools, pip, wheel...done.
Setting project for aws-py-stepfunctions-W_WOZN1B to /Users/luke/go/src/github.com/pulumi/examples/aws-py-stepfunctions

Virtualenv location: /Users/luke/.local/share/virtualenvs/aws-py-stepfunctions-W_WOZN1B
requirements.txt found, instead of Pipfile! Converting...
Warning: Your Pipfile now contains pinned versions, if your requirements.txt did.
We recommend updating your Pipfile to specify the "*" version, instead.

luke:~/go/src/github.com/pulumi/examples/aws-py-stepfunctions (lukehoban/python3)
$ pipenv install
Pipfile.lock not found, creating...
Locking [dev-packages] dependencies...
Locking [packages] dependencies...
lib/python3.7/site-packages/pipenv/utils.py", line 402, in resolve_deps
    req_dir=req_dir
  File "/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/pipenv/utils.py", line 250, in actually_resolve_deps
    req = Requirement.from_line(dep)
  File "/usr/local/Cellar/pipenv/2018.7.1/libexec/lib/python3.7/site-packages/pipenv/vendor/requirementslib/models/requirements.py", line 704, in from_line
    line, extras = _strip_extras(line)
TypeError: 'module' object is not callable
```

Two questions:
1. Is the error here related to our packages somehow?
2. If we are going to recommend moving to `pipenv` should we be moving from `requirements.txt` to `Pipfile` in examples/templates/etc.?

Note that I have not had any problems using `virtualenv`.

swgillespie replied:
> Is the error here related to our packages somehow?
> TypeError: 'module' object is not callable

No, that's a bug in pipenv. Arg... I guess pipenv is still quite buggy. Pipenv 2018.7.1 is quite old, though, they might have fixed that issue already.

Virtualenv is certainly the more stable option, and if you're having trouble with it I'm sure you won't be the last, so I should probably leave the virtualenv instructions in place.

lukehoban commented:
Going to merge this so we have updated docs - we can revisit `pipenv` in the future if needed.
